libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp

```diff
@@ -0,0 +1,277 @@
+
+#include <utility>
+#include <unistd.h>
+#include <AP_HAL_Linux/GPIO.h>
+
+extern const AP_HAL::HAL &hal;
+
+
+////////////////////////////
+// LSM6DS3 Gyro Registers //
+////////////////////////////
+#define WHO_AM_I_ADDR				0x0F
+#define LSM6DS3_WHOAMI  			0x69 
+#define FIFO_CTRL5		            0x0A
+#define CTRL10_C					0x19
+#define INT1_CTRL					0x0d
+#define CTRL2_G						0x11
+#define CTRL1_XL					0x10 
+#define CTRL9_XL					0x18
+#define OUT_X_L_A					0x28
+#define OUT_X_L_G					0x22	
+#define STATUS_REG					0x1E 
+#define XL_DATA_READY_MASK			0x1
+#define G_DATA_READY_MASK           (0x1 << 1)
+#define CTRL1_XL_1600khz            (0x8 << 4)
+#define CTRL1_XL_16g                (0x1 << 2)
+#define CTRL9_XL_AXIS_VALUES		(0x7 << 3) 
+#define CTRL10_C_AXIS_VALUES		(0x7 << 3) 
+
+AP_InertialSensor_LSM6DS3::AP_InertialSensor_LSM6DS3(AP_InertialSensor &imu,AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+    : AP_InertialSensor_Backend(imu)
+	,_dev(std::move(dev))
+{
+}
+
+AP_InertialSensor_Backend *AP_InertialSensor_LSM6DS3::probe(AP_InertialSensor &_imu,AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    if (!dev) 
+	{
+        return nullptr;
+    }
+    AP_InertialSensor_LSM6DS3 *sensor =
+        new AP_InertialSensor_LSM6DS3(_imu, std::move(dev));
+    if (!sensor || !sensor->_init())
+	{
+        delete sensor;
+        return nullptr;
+    }
+    return sensor;
+}
+
+
+
+bool AP_InertialSensor_LSM6DS3::_init()
+{
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        return false;
+    }
+    uint8_t tries, whoami;    
+    whoami = _register_read(WHO_AM_I_ADDR);
+    if (whoami != LSM6DS3_WHOAMI) {	//not LSM6DS3
+        hal.console->printf("LSM6DS3: unexpected acc/mag  WHOAMI 0x%x\n", (unsigned)whoami);
+        _dev->get_semaphore()->give();
+		return false;
+    }
+        
+    for (tries = 0; tries < 5; tries++)
+	{
+        _dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+  		_dev->write_register(FIFO_CTRL5, 0x50); //fifo odr 6.6khz , fifo bypass mode
+		_accel_init();
+		_gyro_init();
+		_dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+		hal.scheduler->delay(10);
+		if (_accel_data_ready() && _gyro_data_ready())
+		{ 
+			break;
+		}
+	}// end tries for
+    if (5 == tries)
+	{
+		hal.console->printf("Failed to boot LSM6DS3 5 times\n\n");
+        _dev->get_semaphore()->give();
+		return false;
+    }
+    _dev->get_semaphore()->give();
+    return true;
+}
+
+
+/*
+  get the sensor going
+ */
+void AP_InertialSensor_LSM6DS3::start(void)
+{
+    _accel_instance = _imu.register_accel(800, _dev->get_bus_id_devtype(DEVTYPE_LSM6DS3));
+    _gyro_instance = _imu.register_gyro(800, _dev->get_bus_id_devtype(DEVTYPE_LSM6DS3));
+
+    // start the timer process to read samples
+    _dev->register_periodic_callback(1250, FUNCTOR_BIND_MEMBER(&AP_InertialSensor_LSM6DS3::_poll_data, void));
+
+}
+
+
+uint8_t AP_InertialSensor_LSM6DS3::_register_read(uint8_t reg)
+{
+    uint8_t val = 0;
+	if (_sem->take_nonblocking()) 
```

Comment:
why are you taking the frontend semaphore for a register read?

Reply:
I'm very sorry, this driver was tested, and it's values were verified to be correct , but in the hit of the moment those semaphores were added without much thought, and without testing, and make no sense at all. This kind of mistake will not be repeated.